src/app/zeko/sequencer/lib/da_layer.ml

```diff
@@ -1,99 +1,247 @@
-  | None ->
-      print_endline "No da contract address provided. Skipping bootstrapping" ;
-      return []
-  | Some da_contract_address ->
-      let payload =
-        Yojson.to_string
-          (`Assoc
-            [ ("address", `String da_contract_address); ("to", `String to_) ] )
-      in
-      let getBatchesCmd =
-        match Sys.getenv "DA_GET_BATCHES" with
-        | Some cmd ->
-            cmd
-        | None ->
-            "exec $cd ../da-layer && npx hardhat run scripts/getBatches.ts"
+let deploy ~da_websocket ~da_private_key ~quorum ~validators =
+  Da_utils.deploy ~da_websocket ~da_private_key ~quorum ~validators
+
+(* List of accounts in the genesis state *)
+module Genesis_state = struct
+  [%%versioned
+  module Stable = struct
+    module V1 = struct
+      type t = Account.Stable.V2.t list [@@deriving yojson]
+
+      let to_latest = Fn.id
+    end
+  end]
+
+  let get (config : Config.t) =
+    let%bind.Deferred.Result data =
+      Da_utils.get_genesis_state ~da_websocket:config.da_websocket
+        ~da_contract_address:config.da_contract_address
+    in
+    let%bind.Deferred.Result data =
+      match Base64.decode data with
+      | Ok data ->
+          return (Ok data)
+      | Error (`Msg e) ->
+          return (Error (Error.of_string e))
+    in
+    let accounts = Binable.of_string (module Stable.Latest) data in
+    return (Ok accounts)
+
+  let init (config : Config.t) ~(genesis_accounts : Account.t list) =
+    let data =
+      Base64.encode_string
+      @@ Binable.to_string (module Stable.Latest) genesis_accounts
+    in
+    Da_utils.init_genesis_state ~da_websocket:config.da_websocket
+      ~da_contract_address:config.da_contract_address
+      ~da_private_key:config.da_private_key ~data
+end
+
+(**
+  Mapping of Account_id.t -> 'a
+  Used for storing the receipt chain hashes of accounts in the batch
+*)
+module Account_map = struct
+  include Core.Map.Make (Account_id)
+
+  let to_yojson a_to_yojson (m : 'a t) =
+    to_alist m |> [%to_yojson: (Account_id.t * 'a) list] a_to_yojson
+
+  let of_yojson a_of_yojson json =
+    let%map.Result alist =
+      [%of_yojson: (Account_id.t * 'a) list] a_of_yojson json
+    in
+    of_alist_exn alist
+end
+
+(**
+  A batch is a collection of user commands that have been applied to the ledger state.
+  The batch also contains the receipt chain hashes of the accounts that are updated by the commands.
+  
+  The receipt chain hashes are used to verify the validity of the batch by the DA validators.
+  The validator applies all the receipts to the chain hashes and compares with the target ledger.
+  If all of them match, they sign the target ledger hash together with the location.
+*)
+module Batch = struct
+  [%%versioned
+  module Stable = struct
+    module V1 = struct
+      type t =
+        { previous_location : Bounded_types.String.Stable.V1.t
+        ; receipt_chain_hashes :
+            (Account_id.Stable.V2.t * Receipt.Chain_hash.Stable.V1.t) list
```

Comment:
Why do you need this?

Reply:
it's explained here https://github.com/zeko-labs/zeko/pull/163/files/784c76e2a8b09266c301fc51ae8340d79dffb81a#diff-37b76c20662619a87a2bcc5cf60d2f4fdabda824fbb26da4098c4e67e3f64d31R83-R85

the validator applies the receipts to these hashes and compares with the target ledger

Reply:
I mean, why do you need a list of receipt chain hashes and the account it's for? You should have all the data you need in the ledger and commands.

Reply:
in the ledger you have only the target receipt chain hashes, you need the source receipt chain hashes

Reply:
technically you could have target ledger without receipt chain hashes, but that would be unnecessarily complicated

Reply:
or we could have source_ledger and fully apply commands, but we've agreed that validators would be just applying receipts

Reply:
What do you mean by receipt?

Reply:
signed command payload hash / zkapp full txn commitment

Reply:
But you can get this from the commands? You want to verify that the commands posted to the DA are correct.

Reply:
But I need the preimage, which is the state of receipt chain hash in the source ledger.

Reply:
But you already have that by reading the old ledger?

Reply:
I have only the target ledger, not the old ledger

Reply:
You should have the old ledger too. It's on the DA layer.

Reply:
That would require the da signer to either maintain the state of the rollup (multiple of them if it forks) or build the state each time. This way the da signer is stateless.

Reply:
It would not need to know the state of the rollup as mentioned in #12. It only needs state of DA. 

Maybe your scheme is secure but I haven’t thought about it and am likely misunderstanding what you mean.

Reply:
In fact even with scheme in #12 you don’t need any expensive state. DA node only needs to write to DA, never read, since they can trust their own signatures.

Reply:
Yes you can